src/object/cli_obj.c

```diff
@@ -1430,37 +1426,60 @@ obj_iod_valid(unsigned int nr, daos_iod_t *iods, bool update)
 		switch (iods[i].iod_type) {
 		default:
 			D_ERROR("Unknown iod type=%d\n", iods[i].iod_type);
-			return false;
+			return -DER_INVAL;
 
 		case DAOS_IOD_NONE:
 			if (!iods[i].iod_recxs && iods[i].iod_nr == 0)
 				continue;
 
 			D_ERROR("IOD_NONE ignores value iod_nr=%d, recx=%p\n",
 				 iods[i].iod_nr, iods[i].iod_recxs);
-			return false;
+			return -DER_INVAL;
 
 		case DAOS_IOD_ARRAY:
 			if (!update && iods[i].iod_nr == 0) /* size query */
 				continue;
 
-			if (obj_recx_valid(iods[i].iod_nr, iods[i].iod_recxs,
+			if (sgls == NULL) {
+				/* size query or punch */
+				if (!update || iods[i].iod_size == 0)
+					continue;
+				D_ERROR("invalid update req with NULL sgl\n");
+				return -DER_INVAL;
+			}
+			if (!obj_recx_valid(iods[i].iod_nr, iods[i].iod_recxs,
 					   update)) {
-				continue;
-			} else {
 				D_ERROR("IOD_ARRAY should have valid recxs\n");
-				return false;
+				return -DER_INVAL;
 			}
+			rc = obj_req_size_valid(daos_iods_len(&iods[i], 1),
+					daos_sgl_buf_size(&sgls[i]), update);
+			if (rc)
+				return rc;
+			break;
 
 		case DAOS_IOD_SINGLE:
-			if (iods[i].iod_nr == 1)
-				continue;
-
-			D_ERROR("IOD_SINGLE iod_nr %d != 1\n", iods[i].iod_nr);
-			return false;
+			if (iods[i].iod_nr != 1) {
+				D_ERROR("IOD_SINGLE iod_nr %d != 1\n",
+					iods[i].iod_nr);
+				return -DER_INVAL;
+			}
+			if (sgls == NULL) {
+				/* size query or punch */
+				if (!update || iods[i].iod_size == 0)
```

Comment:
please use opc here? it would be clearer.

Reply:
as above, only for update or fetch. just that some test code do the punch by using zero iod_size.